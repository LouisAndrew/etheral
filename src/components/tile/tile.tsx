import React from 'react';
import Img, { FluidObject, FixedObject } from 'gatsby-image';

import { Box, Flex, FlexProps, Heading } from 'rebass';

type Props = FlexProps & {
    url: string;
    imgAlt: string;
    img?: {
        sources: FluidObject | FluidObject[] | FixedObject | FixedObject[];
    };
    tileOnText?: string;
    imageIsFluid?: boolean;
};

const Tile: React.FC<Props> = ({
    img,
    tileOnText,
    url,
    imgAlt,
    imageIsFluid,
    css,
    ...rest
}) => {
    console.log(css); // css. rather than dealing with the overhead here, I just decided not to use the css props -> https://github.com/rebassjs/rebass/issues/664

    return (
        <Flex
            data-testid="tile-container"
            alignItems="center"
            justifyContent="center"
            sx={{ position: 'relative' }}
            {...rest}
        >
            <Box
                height="100%"
                width="100%"
                sx={{ position: 'absolute', top: 0, left: 0, zIndex: -1 }}
            >
                {img ? (
                    imageIsFluid ? (
                        <Img
                            fluid={img.sources as FluidObject | FluidObject[]}
                            alt={imgAlt}
                        />
                    ) : (
                        <Img
                            fixed={img.sources as FixedObject | FixedObject[]}
                            alt={imgAlt}
                        />
                    )
                ) : (
                    <img src={url} alt={imgAlt} />
                )}
            </Box>
            {tileOnText && (
                <Heading as="h3" variant="tileText">
                    {tileOnText}
                </Heading>
            )}
        </Flex>
    );
};

export { Tile };
