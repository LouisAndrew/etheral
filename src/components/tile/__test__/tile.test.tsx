import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Tile } from '../tile';

describe('Tile component', () => {
    const mockTextContent = 'Hello, World!';

    const elementWithUrl = <Tile url="" imgAlt="mockAltImage" />;
    const elementExtended = <Tile url="" imgAlt="mockAltImage" width="50%" />;
    const elementWithText = (
        <Tile url="" imgAlt="mockAltImage" tileOnText={mockTextContent} />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(elementWithUrl, div);
    });

    it('should render extended element without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(elementExtended, div);
    });

    it('should render extended element with style rules correctly', () => {
        const { getByTestId } = render(elementExtended);
        const TileContainer = getByTestId('tile-container');

        expect(TileContainer).toHaveStyle('width: 50%;');
    });

    it('should render element with text correctly', () => {
        const { getByRole } = render(elementWithText);

        const Text = getByRole('heading');
        expect(Text).toBeInTheDocument();
        expect(Text).toHaveTextContent(mockTextContent);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(elementWithUrl).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
