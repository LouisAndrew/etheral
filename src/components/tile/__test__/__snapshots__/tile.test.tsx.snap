// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile component matches snapshot 1`] = `
<div
  className="css-12w9nlj"
  data-testid="tile-container"
>
  <div
    className="css-oryowi"
  >
    <img
      alt="mockAltImage"
      src=""
    />
  </div>
</div>
`;
